Matchismo/CardMatchingGameViewController.m

```diff
@@ -0,0 +1,237 @@
+
+@interface CardMatchingGameViewController ()
+
+@property (strong, nonatomic) IBOutletCollection(UIButton) NSArray *cardButtons;
+@property (weak, nonatomic) IBOutlet UILabel *scoreLabel;
+@property (weak, nonatomic) IBOutlet UILabel *infoLabel;
+@property (weak, nonatomic) IBOutlet UISlider *historySlider;
+@property (strong, nonatomic) CardMatchingGame *game;
+@property (strong, nonatomic) NSMutableArray *history; // of CardMatchingGame
+@property (strong, nonatomic) NSMutableArray *historyInfos; // of NSAttributedString
+
+@end
+
+@implementation CardMatchingGameViewController
+
+NSString *const USER_DEFAULTS_HIGH_SCORES = @"HighScores";
+NSString *const HIGH_SCORES_CATEGORY_SCORE = @"HighScoresCategoryScore";
+NSString *const HIGH_SCORES_CATEGORY_TIME = @"HighScoresCategoryTime";
+
+- (void)viewDidLoad
+{
+	self.historySlider.continuous = YES;
+	self.historySlider.minimumValue = 0;
+	[self updateHistory];
+	self.historySlider.enabled = NO;
+}
+
+- (NSMutableArray *)history
+{
+	if (!_history) {
+		_history = [[NSMutableArray alloc] init];
+	}
+	return _history;
+}
+
+- (NSMutableArray *)historyInfos
+{
+	if (!_historyInfos) {
+		_historyInfos = [[NSMutableArray alloc] init];
+	}
+	return _historyInfos;
+}
+
+- (CardMatchingGame *)game
+{
+	if (!_game) {
+		_game = [self createGame];
+		_game.mode = [self gameMode];
+	}
+	return _game;
+}
+
+
++ (NSString *)gameName // abstract
+{
+	return nil;
+}
+
+- (GameMode)gameMode // abstract
+{
+	return -1;
+}
+
+- (Deck *)createDeck // abstract
+{
+	return nil;
+}
+
+- (CardMatchingGame *)createGame
+{
+	return [[CardMatchingGame alloc] initWithCardCount:[self.cardButtons count]
+											 usingDeck:[self createDeck]];
+}
+
+- (void)updateHistory
+{
+	[self.history addObject:[self.game copy]];
+	self.historySlider.enabled = YES;
+	self.historySlider.maximumValue = [self.history count] - 1;
+	if (self.historySlider.maximumValue == 0) {
+		self.historySlider.maximumValue = 1;
+	}
+	self.historySlider.value = self.historySlider.maximumValue;
+}
+
+- (IBAction)touchCardButton:(UIButton *)sender
+{
+	if (self.historySlider.value != self.historySlider.maximumValue) {
+		return;
+	}
+
+	NSUInteger chosenButtonIndex = [self.cardButtons indexOfObject:sender];
+	[self.game chooseCardAtIndex:chosenButtonIndex];
+	[self updateUI];
+	[self updateHistory];
+	if ([self.game isEnded]) {
+		[self saveGame];
+	}
+}
+
+- (void)saveGame
+{
+	NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+
+	NSMutableDictionary *highScores = [[userDefaults dictionaryForKey:(NSString *)USER_DEFAULTS_HIGH_SCORES] mutableCopy];
+	if (!highScores) {
+		highScores = [[NSMutableDictionary alloc] init];
+	}
+
+	NSMutableDictionary *gameHighScores = [[highScores objectForKey:[[self class] gameName]] mutableCopy];
+	if(!gameHighScores) {
+		gameHighScores = [[NSMutableDictionary alloc] init];
+		[gameHighScores setObject:[[NSMutableArray alloc] init] forKey:HIGH_SCORES_CATEGORY_SCORE];
```

Comment:
You could use subscripting syntax.
* [Rypress, NSDictionary](http://rypress.com/tutorials/objective-c/data-types/nsdictionary) (Chapter: Adding and Removing Entries)
* [NSHipster, Object Subscripting](http://nshipster.com/object-subscripting/)